orchestrator-salt/src/main/resources/salt/salt/postgresql/upgrade/scripts/backup_db.sh

```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env bash
+# backup_db.sh
+# This script is needed as Azure does not support in place RDS upgrade
+# This script:
+#  - creates backup local temp directory
+#  - dumps the global objects and roles to a SQL file with `psql`
+#  - for each service:
+#     - backs up service data in a parallel manner with `pg_dump` to the temp directory
+
+set -o nounset
+set -o pipefail
+
+if [[ $# -ne 3 || "$1" == "None" ]]; then
+  echo "Invalid inputs provided"
+  echo "Script accepts 3 inputs:"
+  echo "  1. PostgreSQL host name."
+  echo "  2. PostgreSQL port."
+  echo "  3. PostgreSQL user name."
+  exit 1
+fi
+
+source /opt/salt/scripts/common_utils.sh
+
+HOST="$1"
+PORT="$2"
+USERNAME="$3"
+LOGFILE=/var/log/postgres_upgrade_backup.log
+echo "Logs at ${LOGFILE}"
+
+exec 3>&1 4>&2
+trap 'exec 2>&4 1>&3' 0 1 2 3
+exec 1> >(tee -a "${LOGFILE}") 2> >(tee -a "${LOGFILE}" >&2)
+
+{%- from 'postgresql/settings.sls' import postgresql with context %}
+{% if postgresql.ssl_enabled == True %}
+export PGSSLROOTCERT="{{ postgresql.root_certs_file }}"
+export PGSSLMODE=verify-full
+{%- endif %}
+
+dump_global_objects() {
+  GLOBAL_OBJECT_BACKUP=${DATE_DIR}/roles.sql
+  doLog "INFO Dumping the global objects from database to ${GLOBAL_OBJECT_BACKUP}"
+  pg_dumpall -r --host="$HOST" --port="$PORT" --username="$USERNAME" --database=postgres > $GLOBAL_OBJECT_BACKUP 2> >(tee -a $LOGFILE >&2) || errorExit "Unable to dump global objects"
+}
+
+backup_database_for_services() {
+  {% for service, values in pillar.get('postgres', {}).items()  %}
+  {% if values['user'] is defined %}
+  SERVICE="{{ service }}"
+
+  is_database_exists $SERVICE
+  if [ "$?" -eq 1 ];then
+    doLog "WARN database for $SERVICE doesn't exists"
+    return 0
+  fi
+
+  limit_incomming_connection $SERVICE 0
+  close_existing_connections $SERVICE
+
+  doLog "INFO Dumping ${SERVICE} ..."
+
+  LOCAL_BACKUP=${DATE_DIR}/${SERVICE}_backup
+  pg_dump -Fd -v  --host="$HOST" --port="$PORT" --username="$USERNAME" --dbname=$SERVICE -j 8 -f $LOCAL_BACKUP > >(tee -a $LOGFILE) 2> >(tee -a $LOGFILE >&2) || errorExit "Unable to dump ${SERVICE}"
```

Comment:
Pls check here: https://stackoverflow.com/questions/363223/how-do-i-get-both-stdout-and-stderr-to-go-to-the-terminal-and-a-log-file

Reply:
my question is still valid. as you are redirecting both stdout and stderr to the same logfile, the double tee seems to be an extra step. Or would you like to keep error output on the stderr for some reason?

Reply:
I like this:

The second tee process has it's stdout redirected to the caller's stderr (which is good, because it's stdin is reading from the_cmd's stderr). So when it writes to its stdout, those bits go to the caller's stderr.